### PR TITLE
Expand testing docs

### PR DESCRIPTION
**Pull request by vanitabarrett**
Closes https://github.com/alphagov/govuk-frontend/issues/2721

Expand our existing testing documentation to talk about the different types of testing we do and when to use them.
Make reference to unit testing, which we're introducing for JavaScript logic.

### CODE REVIEW

File: docs/releasing/testing-and-linting.md [hunk 1]
```diff
@@ -2,45 +2,80 @@
...
 
 See [JavaScript Coding Standards](/docs/contributing/coding-standards/js.md#formatting-and-linting) for details.
 
 ## Unit and functional tests with Node.js
 
 We use [Jest](https://jestjs.io/), an automated testing platform with an assertion library, and [Puppeteer](https://pptr.dev/) that is used to control [headless Chrome](https://developers.google.com/web/updates/2017/04/headless-chrome).
 
-We test individual components and for instance global Sass files. See [all.test.js](../../src/govuk/all.test.js). for examples of global tests we run, and  [checkboxes.test.js](../../src/govuk/components/checkboxes/checkboxes.test.js) for an example of component tests.
+### Component tests
+We write tests for every component to check the output of our Nunjucks code. These are found in `template.test.js` files in each component directory. These Nunjucks tests call `render` on the component examples defined in the component yaml files, and asserts the HTML tags, attributes and classes are as expected. For example: checking that when you pass in an `id` to the component using the Nunjucks macro, it outputs the component with an `id` attribute equal to that value.
```
Comment (owenatgov) [inline docs/releasing/testing-and-linting.md:61]: What type of test would component tests be? Unit or functional? I'm wondering if we should sus this out and jot it up as a conscious choice. I _think_ this is functional as we're doing less logic based tests compared to making sure stuff renders properly.

File: docs/releasing/testing-and-linting.md [hunk 2]
```diff
@@ -2,45 +2,80 @@
...
 
-We aim to write the description of our tests in as "natural language" as possible, for instance "back-link component fails to render if the required fields are not included".
+If a component uses JavaScript, we also write tests in a `[component name].test.js` file, for example [checkboxes.test.js](../../src/govuk/components/checkboxes/checkboxes.test.js). These component tests check that interactions, e.g: a mouse click, have the expected result, e.g: the component gaining focus or the addition of a class.
 
-### Running individual tests
-You can run a subset of the test suite that only tests components by running:
+Component JavaScript logic should also be tested using unit tests, in a `[component name].unit.test.mjs` file.
```
Comment (owenatgov) [inline docs/releasing/testing-and-linting.md:65]: Once we have a unit test, we should provide an example as we have done on line 63.

**Review by owenatgov — COMMENTED** [review]
I really like this. Being explicit about how we test stuff at worst can't hurt but more likely is going to help us stay consistent and sensible. I've left a couple of thoughts.

Once this is in a more cohesive state it's worth running this past a tech writer.

File: docs/releasing/testing-and-linting.md (hunk 1 above)
Comment (vanitabarrett) [reply docs/releasing/testing-and-linting.md:61]: I think functional too. I'll add that in here somewhere...

**Comment by vanitabarrett** [comment]
@claireashworth It would be great to get your thoughts and comments on this when you have time (it's not urgent)! Give me a shout if you want me to run you through it too.

File: docs/contributing/coding-standards/components.md [hunk 3]
```diff
@@ -21,7 +21,8 @@ When creating your component, you should create the following files in the compo
 If your component uses JavaScript, you must also create the following files in the component’s folder:
 
       - `_[component-name].mjs`
-      - `_[component-name].test.js`
+      - `[component-name].unit.test.mjs`
```
Comment (claireashworth) [inline docs/contributing/coding-standards/components.md:24]: Just checking that the underscores aren't needed before the square brackets here.

File: docs/contributing/testing.md [hunk 4]
```diff
@@ -65,27 +65,36 @@ You should test that your contribution works:
 
 ## 5. Run the automated tests
 
+Automated testing helps ensure that our code is as free of bugs as it can be. It also lets us know immediately if a change we make breaks any other part of the code.
```
Comment (claireashworth) [inline docs/contributing/testing.md:68]: Should we add how it lets us know? Perhaps - It also provides immediate notifications if a change we make...etc
or
It also immediately displays where a change we make...etc

File: docs/contributing/testing.md [hunk 5]
```diff
@@ -65,27 +65,36 @@ You should test that your contribution works:
...
+
 In your project folder, run `npm test` to run the automated tests, including linting.
 
 If a test fails, you should check your code for any errors, then update any tests you need to.
```
Comment (claireashworth) [inline docs/contributing/testing.md:72]: are they only updating the tests that failed?

Comment (claireashworth) [inline docs/contributing/testing.md:72]: Are they only updating the tests that fail?

File: docs/contributing/testing.md [hunk 6]
```diff
@@ -65,27 +65,36 @@ You should test that your contribution works:
...
+You can [read more about the different types of tests in this project](../releasing/testing-and-linting).
+
 You should write new tests if you’ve created a new component, or changed the way a component works by:
 
 - changing or adding to the component's JavaScript code
 - changing or adding to the component's Nunjucks macro
 - creating or updating a Sass mixin or function
 
-Test files use examples from each component’s `.yaml` file, for example `src/govuk/components/button/button.yaml`. When you add or update tests, you can use the existing examples or add new ones.
+If you're new to testing, see existing test files for examples of things to do. Don't let the tests keep you from submitting your contribution! If you're not sure how to do this or are having trouble, submit your pull request anyway. We will help you create the tests and sort out any kind of problem during code review.
```
Comment (claireashworth) [inline docs/contributing/testing.md:84]: Some minor things:
- Change 'Don't' to 'Do not'  
- "If you're not sure how to do this..." Is 'this' how to do the pull request or the testing?  
- Alternative text tweak for the last line - 'We will help you create the tests and solve problems during code review.'

File: docs/releasing/testing-and-linting.md [hunk 7]
```diff
@@ -2,55 +2,88 @@
 
 The CI lints SASS and JavaScript, runs unit and functional tests with Node, and generates snapshots for [visual regression testing](https://www.browserstack.com/percy/visual-testing).
 
-## Running all tests locally
+## Testing terminology
+
+We use different types of tests to check different areas of our code are working as expected.
+
+Unit tests are small, modular tests that verify a "unit" of code. We write unit tests to check our JavaScript logic, particularly 'background logic' that doesn't heavily rely on DOM interaction (where functional tests may be better suited).
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:9]: Expand 'doesn't' into 'does not'.
Also, does DOM need to be expanded?

File: docs/releasing/testing-and-linting.md [hunk 8]
```diff
@@ -2,55 +2,88 @@
...
+
+Functional tests verify the output of an action and do not check the intermediate states of the system when performing that action. We write functional tests to check component interactions have the expected results, so we also refer to these as component tests. We also write functional tests to check that our Nunjucks code outputs expected HTML, and we tend to refer to these as our Nunjucks tests.
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:11]: Suggest removing 'tend to'. If we use more than one terminology for the test, we should say under what circumstances we call them Nunjucks tests or [some other tests] under other circumstances.

File: docs/releasing/testing-and-linting.md [hunk 9]
```diff
@@ -2,55 +2,88 @@
...
+### Component tests
+We write functional tests for every component to check the output of our Nunjucks code. These are found in `template.test.js` files in each component directory. These Nunjucks tests call `render` on the component examples defined in the component yaml files, and asserts the HTML tags, attributes and classes are as expected. For example: checking that when you pass in an `id` to the component using the Nunjucks macro, it outputs the component with an `id` attribute equal to that value.
 
-We aim to write the description of our tests in as "natural language" as possible, for instance "back-link component fails to render if the required fields are not included".
+If a component uses JavaScript, we also write functional tests in a `[component name].test.js` file, for example [checkboxes.test.js](../../src/govuk/components/checkboxes/checkboxes.test.js). These component tests check that interactions, e.g: a mouse click, have the expected result, e.g: the component gaining focus or the addition of a class/HTML attribute.
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:63]: Remove e.g. - These component tests check that interactions, such as a mouse click, have the expected result. 
I don't think the last example (the component gaining focus) is needed.

File: docs/releasing/testing-and-linting.md [hunk 10]
```diff
@@ -2,55 +2,88 @@
...
 
-### Running individual tests
-You can run a subset of the test suite that only tests components by running:
+Component JavaScript logic should also be tested using unit tests, in a `[component name].unit.test.mjs` file. These tests are better suited for testing behind-the-scenes logic or in cases where the final output of some logic isn't a change to the component markup.
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:65]: Suggest removing the passive voice by changing to - 'You should also test Component Javascript logic with unit tests...etc' 
Also expand 'isn't' to 'is not'. (this is only because the style guide says not to contract negative statements)
Add a comma after 'logic'.

File: docs/releasing/testing-and-linting.md [hunk 11]
```diff
@@ -2,55 +2,88 @@
...
 
-    `npm test -- src/govuk/components/button`
+### Global tests
+We write functional tests for checking our JavaScript exports and our global sass variables - see [all.test.js](../../src/govuk/all.test.js) for examples of global tests we run.
 
-Note: There's a watch mode that keeps a testing session open waiting for changes that can be used with:
+### Conventions
+We aim to write the description of our tests in as "natural language" as possible, for instance "back-link component fails to render if the required fields are not included".
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:71]: Would changing 'natural language' still work if it was changed to 'plain English'? Something like this - We aim to write the test descriptions in plain English. For example, "back-link...etc

File: docs/releasing/testing-and-linting.md [hunk 12]
```diff
@@ -2,55 +2,88 @@
...
 
-    `npm test -- --watch src/govuk/components/button`
+Each test should be independent and isolated from all other tests. It shouldn't matter what order tests are run in, or how many tests from a test suite are run.
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:73]: Suggest changing to something like - Keep all tests separate from each other. It should not matter the order or amount of tests you run from a test suite.

File: docs/releasing/testing-and-linting.md [hunk 13]
```diff
@@ -2,55 +2,88 @@
...
+
+Try and keep assertions small, so they only check for one thing. This makes tests more readable and makes it easier to see what's happening if a test is failing.
```
Comment (claireashworth) [inline docs/releasing/testing-and-linting.md:75]: amend to ...so each test only checks for one thing.

File: docs/contributing/coding-standards/components.md (hunk 3 above)
Comment (vanitabarrett) [reply docs/contributing/coding-standards/components.md:24]: The underscores aren't needed - we don't use underscores in these filenames. I've actually just realised we don't use them in the `.mjs` files either (the top bullet), so I'll remove that too!

File: docs/contributing/testing.md (hunk 5 above)
Comment (vanitabarrett) [reply docs/contributing/testing.md:72]: I think there probably is a scenario where someone may need to update tests that *didn't* fail. They might spot some tests which should have failed, but didn't - they should also update those to make sure the tests are specific enough that they will catch code changes in future.

File: docs/releasing/testing-and-linting.md (hunk 11 above)
Comment (vanitabarrett) [reply docs/releasing/testing-and-linting.md:71]: I'm not sure - do you think that has the same meaning? I think we're trying to say that test names should make sense as a sentence when read out loud, which may be slightly different to plain English. Although I'm struggling to think of an example 🤔

**Comment by vanitabarrett** [comment]
Thanks @claireashworth - I think I've made all your suggested changes now, apart from two comments I wasn't sure about, where I've replied to your comments instead. You should be able to see my replies in the comment threads above this message.

File: docs/contributing/testing.md (hunk 5 above)
Comment (claireashworth) [reply docs/contributing/testing.md:73]: Well of we get any feedback that people are not sure what we mean then we can revisit.

File: docs/contributing/testing.md [hunk 14]
```diff
@@ -65,27 +65,37 @@ You should test that your contribution works:
 
 ## 5. Run the automated tests
 
+Automated testing helps ensure that the code changes we make don't unintentionally break functionality.
```
Comment (claireashworth) [inline docs/contributing/testing.md:68]:
```suggestion
Automated testing helps ensure that the code changes we make do not unintentionally break functionality.
```

File: docs/releasing/testing-and-linting.md (hunk 11 above)
Comment (claireashworth) [reply docs/releasing/testing-and-linting.md:71]: Ah I see. Well 'colloquial' would be an accurate word, but not particularly user friendly. How about 'everyday language' or 'conversational language'?

Comment (vanitabarrett) [reply docs/releasing/testing-and-linting.md:71]: @claireashworth Silly question - how would you word that as a sentence? E.g: "we aim to write the test descriptions in everyday language" - would that be ok?

Comment (claireashworth) [reply docs/releasing/testing-and-linting.md:71]: yup that sounds good

File: docs/releasing/testing-and-linting.md [hunk 15]
```diff
@@ -2,55 +2,88 @@
...
-This will trigger [standard](https://github.com/standard/standard) and [stylelint](https://github.com/stylelint/stylelint) for linting, and [Jest](https://github.com/facebook/jest) for unit and functional tests.
+This will trigger [standard](https://github.com/standard/standard) and [stylelint](https://github.com/stylelint/stylelint) for linting, and [Jest](https://github.com/facebook/jest) for our testing.
 
 See [Tasks](tasks.md) for details of what `npm test` does.
 
-## SASS linting
+### Running individual tests
+You can run a subset of the test suite that only tests components by running:
+
+    `npm test -- --testPathPattern=src/govuk/components/button`
```
Comment (36degrees) [inline docs/releasing/testing-and-linting.md:34]: Interesting! I didn't know you could do it this way. I use `npx jest src/govuk/components/button` which _might_ be easier to remember 🤔

File: docs/releasing/testing-and-linting.md [hunk 16]
```diff
@@ -2,55 +2,88 @@
...
 
 See [JavaScript Coding Standards](/docs/contributing/coding-standards/js.md#formatting-and-linting) for details.
 
 ## Unit and functional tests with Node.js
 
 We use [Jest](https://jestjs.io/), an automated testing platform with an assertion library, and [Puppeteer](https://pptr.dev/) that is used to control [headless Chrome](https://developers.google.com/web/updates/2017/04/headless-chrome).
 
-We test individual components and for instance global Sass files. See [all.test.js](../../src/govuk/all.test.js). for examples of global tests we run, and  [checkboxes.test.js](../../src/govuk/components/checkboxes/checkboxes.test.js) for an example of component tests.
+### Component tests
+We write functional tests for every component to check the output of our Nunjucks code. These are found in `template.test.js` files in each component directory. These Nunjucks tests call `render` on the component examples defined in the component yaml files, and asserts the HTML tags, attributes and classes are as expected. For example: checking that when you pass in an `id` to the component using the Nunjucks macro, it outputs the component with an `id` attribute equal to that value.
```
Comment (36degrees) [inline docs/releasing/testing-and-linting.md:61]: Minor suggestion for readability

```suggestion
We write functional tests for every component to check the output of our Nunjucks code. These are found in `template.test.js` files in each component directory. These Nunjucks tests render the component examples defined in the component yaml files, and assert that the HTML tags, attributes and classes are as expected. For example: checking that when you pass in an `id` to the component using the Nunjucks macro, it outputs the component with an `id` attribute equal to that value.
```

**Review by 36degrees — APPROVED** [review]
This looks great, thanks @vanitabarrett! I have a couple of minor comments but nothing blocking.

File: docs/releasing/testing-and-linting.md (hunk 15 above)
Comment (vanitabarrett) [reply docs/releasing/testing-and-linting.md]: Nice, didn't know about that 👍 I'll update this bit of the docs.
